"Debug Failure" with misplaced 'continue' and async/await

If the `continue` keyword is used outside of a loop, and there's an `await` on the other side of the conditional (see sample code below), this causes an ugly error.

**TypeScript Version:**  2.4.2

**Code**

```ts
function dummyPromise() {
    return new Promise((resolve, reject) => {});
}

async function func() {
    if (true) {
        continue;
    } else {
        await dummyPromise();
    }
}
```

**Expected behavior:**

This should print `error TS1107: Jump target cannot cross function boundary.` (this is usually the message when `continue` is wrongly used outside of a loop)

**Actual behavior:**

```
afischer-mba ~/ts-bug-report $ tsc -v
Version 2.4.2
afischer-mba ~/ts-bug-report $ tsc src/test.ts 
/usr/local/lib/node_modules/typescript/lib/tsc.js:55776
                throw e;
                ^

Error: Debug Failure. 
    at findContinueTarget (/usr/local/lib/node_modules/typescript/lib/tsc.js:48265:22)
    at transformAndEmitContinueStatement (/usr/local/lib/node_modules/typescript/lib/tsc.js:47796:25)
    at transformAndEmitStatementWorker (/usr/local/lib/node_modules/typescript/lib/tsc.js:47551:28)
    at transformAndEmitStatement (/usr/local/lib/node_modules/typescript/lib/tsc.js:47531:13)
    at transformAndEmitStatements (/usr/local/lib/node_modules/typescript/lib/tsc.js:47515:17)
    at transformAndEmitEmbeddedStatement (/usr/local/lib/node_modules/typescript/lib/tsc.js:47520:17)
    at transformAndEmitIfStatement (/usr/local/lib/node_modules/typescript/lib/tsc.js:47617:21)
    at transformAndEmitStatementWorker (/usr/local/lib/node_modules/typescript/lib/tsc.js:47541:28)
    at transformAndEmitStatement (/usr/local/lib/node_modules/typescript/lib/tsc.js:47531:13)
    at transformAndEmitStatements (/usr/local/lib/node_modules/typescript/lib/tsc.js:47515:17)
```

--- a/src/compiler/transformers/generators.ts
+++ b/src/compiler/transformers/generators.ts
@@ -1639,8 +1639,13 @@ namespace ts {
 
         function transformAndEmitContinueStatement(node: ContinueStatement): void {
             const label = findContinueTarget(node.label ? unescapeLeadingUnderscores(node.label.escapedText) : undefined);
-            Debug.assert(label > 0, "Expected continue statment to point to a valid Label.");
-            emitBreak(label, /*location*/ node);
+            if (label > 0) {
+                emitBreak(label, /*location*/ node);
+            }
+            else {
+                // invalid continue without a containing loop. Leave the node as is, per #17875.
+                emitStatement(node);
+            }
         }
 
         function visitContinueStatement(node: ContinueStatement): Statement {
@@ -1656,8 +1661,13 @@ namespace ts {
 
         function transformAndEmitBreakStatement(node: BreakStatement): void {
             const label = findBreakTarget(node.label ? unescapeLeadingUnderscores(node.label.escapedText) : undefined);
-            Debug.assert(label > 0, "Expected break statment to point to a valid Label.");
-            emitBreak(label, /*location*/ node);
+            if (label > 0) {
+                emitBreak(label, /*location*/ node);
+            }
+            else {
+                // invalid break without a containing loop, switch, or labeled statement. Leave the node as is, per #17875.
+                emitStatement(node);
+            }
         }
 
         function visitBreakStatement(node: BreakStatement): Statement {
@@ -2351,27 +2361,27 @@ namespace ts {
          * @param labelText An optional name of a containing labeled statement.
          */
         function findBreakTarget(labelText?: string): Label {
-            Debug.assert(blocks !== undefined);
-            if (labelText) {
-                for (let i = blockStack.length - 1; i >= 0; i--) {
-                    const block = blockStack[i];
-                    if (supportsLabeledBreakOrContinue(block) && block.labelText === labelText) {
-                        return block.breakLabel;
+            if (blockStack) {
+                if (labelText) {
+                    for (let i = blockStack.length - 1; i >= 0; i--) {
+                        const block = blockStack[i];
+                        if (supportsLabeledBreakOrContinue(block) && block.labelText === labelText) {
+                            return block.breakLabel;
+                        }
+                        else if (supportsUnlabeledBreak(block) && hasImmediateContainingLabeledBlock(labelText, i - 1)) {
+                            return block.breakLabel;
+                        }
                     }
-                    else if (supportsUnlabeledBreak(block) && hasImmediateContainingLabeledBlock(labelText, i - 1)) {
-                        return block.breakLabel;
+                }
+                else {
+                    for (let i = blockStack.length - 1; i >= 0; i--) {
+                        const block = blockStack[i];
+                        if (supportsUnlabeledBreak(block)) {
+                            return block.breakLabel;
+                        }
                     }
                 }
             }
-            else {
-                for (let i = blockStack.length - 1; i >= 0; i--) {
-                    const block = blockStack[i];
-                    if (supportsUnlabeledBreak(block)) {
-                        return block.breakLabel;
-                    }
-                }
-            }
-
             return 0;
         }
 
@@ -2381,24 +2391,24 @@ namespace ts {
          * @param labelText An optional name of a containing labeled statement.
          */
         function findContinueTarget(labelText?: string): Label {
-            Debug.assert(blocks !== undefined);
-            if (labelText) {
-                for (let i = blockStack.length - 1; i >= 0; i--) {
-                    const block = blockStack[i];
-                    if (supportsUnlabeledContinue(block) && hasImmediateContainingLabeledBlock(labelText, i - 1)) {
-                        return block.continueLabel;
+            if (blockStack) {
+                if (labelText) {
+                    for (let i = blockStack.length - 1; i >= 0; i--) {
+                        const block = blockStack[i];
+                        if (supportsUnlabeledContinue(block) && hasImmediateContainingLabeledBlock(labelText, i - 1)) {
+                            return block.continueLabel;
+                        }
+                    }
+                }
+                else {
+                    for (let i = blockStack.length - 1; i >= 0; i--) {
+                        const block = blockStack[i];
+                        if (supportsUnlabeledContinue(block)) {
+                            return block.continueLabel;
+                        }
                     }
                 }
             }
-            else {
-                for (let i = blockStack.length - 1; i >= 0; i--) {
-                    const block = blockStack[i];
-                    if (supportsUnlabeledContinue(block)) {
-                        return block.continueLabel;
-                    }
-                }
-            }
-
             return 0;
         }
 

--- a/tests/baselines/reference/invalidContinueInDownlevelAsync.errors.txt
+++ b/tests/baselines/reference/invalidContinueInDownlevelAsync.errors.txt
@@ -1,0 +1,17 @@
+tests/cases/compiler/invalidContinueInDownlevelAsync.ts(3,9): error TS1107: Jump target cannot cross function boundary.
+tests/cases/compiler/invalidContinueInDownlevelAsync.ts(6,9): error TS7027: Unreachable code detected.
+
+
+==== tests/cases/compiler/invalidContinueInDownlevelAsync.ts (2 errors) ====
+    async function func() {
+        if (true) {
+            continue;
+            ~~~~~~~~~
+!!! error TS1107: Jump target cannot cross function boundary.
+        }
+        else {
+            await 1;
+            ~~~~~
+!!! error TS7027: Unreachable code detected.
+        }
+    }

--- a/tests/baselines/reference/invalidContinueInDownlevelAsync.js
+++ b/tests/baselines/reference/invalidContinueInDownlevelAsync.js
@@ -1,0 +1,63 @@
+//// [invalidContinueInDownlevelAsync.ts]
+async function func() {
+    if (true) {
+        continue;
+    }
+    else {
+        await 1;
+    }
+}
+
+//// [invalidContinueInDownlevelAsync.js]
+var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
+    return new (P || (P = Promise))(function (resolve, reject) {
+        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
+        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
+        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
+        step((generator = generator.apply(thisArg, _arguments || [])).next());
+    });
+};
+var __generator = (this && this.__generator) || function (thisArg, body) {
+    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
+    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
+    function verb(n) { return function (v) { return step([n, v]); }; }
+    function step(op) {
+        if (f) throw new TypeError("Generator is already executing.");
+        while (_) try {
+            if (f = 1, y && (t = y[op[0] & 2 ? "return" : op[0] ? "throw" : "next"]) && !(t = t.call(y, op[1])).done) return t;
+            if (y = 0, t) op = [0, t.value];
+            switch (op[0]) {
+                case 0: case 1: t = op; break;
+                case 4: _.label++; return { value: op[1], done: false };
+                case 5: _.label++; y = op[1]; op = [0]; continue;
+                case 7: op = _.ops.pop(); _.trys.pop(); continue;
+                default:
+                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
+                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
+                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
+                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
+                    if (t[2]) _.ops.pop();
+                    _.trys.pop(); continue;
+            }
+            op = body.call(thisArg, _);
+        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
+        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
+    }
+};
+function func() {
+    return __awaiter(this, void 0, void 0, function () {
+        return __generator(this, function (_a) {
+            switch (_a.label) {
+                case 0:
+                    if (!true) return [3 /*break*/, 1];
+                    continue;
+                    return [3 /*break*/, 3];
+                case 1: return [4 /*yield*/, 1];
+                case 2:
+                    _a.sent();
+                    _a.label = 3;
+                case 3: return [2 /*return*/];
+            }
+        });
+    });
+}

--- a/tests/cases/compiler/invalidContinueInDownlevelAsync.ts
+++ b/tests/cases/compiler/invalidContinueInDownlevelAsync.ts
@@ -1,0 +1,8 @@
+async function func() {
+    if (true) {
+        continue;
+    }
+    else {
+        await 1;
+    }
+}

